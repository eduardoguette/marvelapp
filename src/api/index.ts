export * from './characters'
export * from './config'

