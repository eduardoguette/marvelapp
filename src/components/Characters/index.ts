export * from './Action'
export * from './Card'
export * from './LikeCounter'
export * from './List'
