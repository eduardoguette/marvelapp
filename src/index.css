:root {
  /* Colores */
  --primary: #ec1d24;
  --secondary: #000000;
  --gray: #333333;
  --white: #ffffff;

  /* fonts */
  --font-basic: Roboto Condense, system-ui, -apple-system, BlinkMacSystemFont,
    'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue',
    sans-serif;

  /* body font */
  --body: normal 16px / 24px var(--font-basic);

  /* headline fonts */
  --headline1: normal 700 40px / 46.88px var(--font-basic);
  --headline2: normal 700 32px / 37.5px var(--font-basic);
  --headline3: normal 700 24px / 28.13px var(--font-basic);
  --headline4: normal 700 16px / 24px var(--font-basic);

  /* caption font */
  --caption: normal 12px / 14.06px var(--font-basic);
}

@font-face {
  font-family: Roboto Condense;
  src: url('./assets/fonts/RobotoCondensed-Regular.ttf') format('truetype');
  font-weight: normal;
  font-style: normal;
}

@font-face {
  font-family: Roboto Condense;
  src: url('./assets/fonts/RobotoCondensed-Bold.ttf') format('truetype');
  font-weight: 700;
  font-style: normal;
}

*,
*::after,
*::before {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
  font: var(--body);
}

li {
  list-style: none;
}
