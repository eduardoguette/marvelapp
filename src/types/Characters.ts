export interface Characters {
  data: CharactersData
}

export interface CharactersData {
  offset: number
  limit: number
  total: number
  count: number
  results: CharactersResult[]
}

export interface CharactersResult {
  id: number
  name: string
  description: string
  modified: string
  thumbnail: Thumbnail
  comics: Comics
}

export interface Comics {
  available: number
  collectionURI: string
  items: ComicsItem[]
}

export interface ComicsItem {
  resourceURI: string
  name: string
}

export interface Thumbnail {
  path: string
  extension: string
}
