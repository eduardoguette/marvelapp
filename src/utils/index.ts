export * from './saveInLocal';
export * from './updateFavorites';
