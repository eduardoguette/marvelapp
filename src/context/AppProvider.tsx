import { CharactersResult } from '@/types'
import { saveInLocal, updateFavorites } from '@/utils'

import { ReactNode, createContext, useEffect, useState } from 'react'

interface IProvider {
  data: null | CharactersResult[]
  filter: string
  setFiler: (filter: string) => void
  favorites: number[]
  addFavorites: (id: number) => void
  setData: (data: CharactersResult[]) => void
}

export const ContextAppProvider = createContext<IProvider>({
  data: null,
  filter: '',
  setFiler: () => {},
  setData: () => {},
  favorites: [],
  addFavorites: () => {}
})

interface AppProvider {
  children: ReactNode
}


export const AppProvider: React.FC<AppProvider> = ({ children }) => {
  const [filter, setFiler] = useState('')
  const [favorites, setFavorites] = useState<number[]>([])

  useEffect(() => {
    setFavorites(saveInLocal.get('heroes') || [])
  }, [])

  const [data, setData] = useState<CharactersResult[] | null>(null)

  const addFavorites = (id: number) => {
    const newFavorites = updateFavorites(favorites, id)
    setFavorites(newFavorites)
    saveInLocal.set('heroes', newFavorites)
  }
  return (
    <ContextAppProvider.Provider
      value={{
        favorites,
        data,
        setData,
        filter,
        setFiler,
        addFavorites
      }}
    >
      {children}
    </ContextAppProvider.Provider>
  )
}
