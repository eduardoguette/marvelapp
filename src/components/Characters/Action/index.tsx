import heart from '@/assets/heart-black.svg'
import heartRed from '@/assets/heart.svg'
import { ContextAppProvider } from '@/context'
import { FC, useContext } from 'react'
import styles from '../characters.module.css'
interface IActionProps {
  idCharacter: number
  characterDetails?: boolean
}
export const Action: FC<IActionProps> = ({ idCharacter, characterDetails }) => {
  const { favorites, addFavorites } = useContext(ContextAppProvider)
  return (
    <button
      className={styles.button}
      onClick={() => {
        addFavorites(idCharacter)
      }}
    >
      {favorites.find((id) => id === idCharacter) ? (
        <img
          src={heartRed}
          alt="heart icon"
          height={characterDetails ? 24 : 12}
          width={characterDetails ? 24 : 12}
        />
      ) : (
        <img
          src={heart}
          alt="heart icon"
          height={characterDetails ? 24 : 12}
          width={characterDetails ? 24 : 12}
        />
      )}
    </button>
  )
}
