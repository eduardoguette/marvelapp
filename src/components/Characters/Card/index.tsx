import { CharactersResult } from '@/types'
import { FC, useEffect } from 'react'
import { Link } from 'react-router-dom'

import { Action } from '../Action'
import styles from '../characters.module.css'
interface ICharacterCard {
  character: CharactersResult
}

export const Card: FC<ICharacterCard> = ({ character }) => {
  useEffect(() => {
    if ('caches' in window) {
      caches.open('character-images').then((cache) => {
        cache.add(
          `${character.thumbnail.path}.${character.thumbnail.extension}`
        )
      })
    }
  }, [character])
  return (
    <li key={character.id} className={styles.card}>
      <Link to={`character/${character.id}`} className={styles.cardLink}>
        <header>
          <img
            src={`${character.thumbnail.path}.${character.thumbnail.extension}`}
            alt={character.description}
            className={styles.cardImage}
          />
        </header>
      </Link>
      <footer className={styles.cardFooter}>
        <p className="">{character.name}</p>
        <Action idCharacter={character.id} />
      </footer>
    </li>
  )
}
