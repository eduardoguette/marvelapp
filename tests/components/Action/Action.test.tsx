import { fireEvent, render, screen } from '@testing-library/react'
import React, { useState } from 'react'
import { describe, expect, it } from 'vitest'
import { Action } from '../../../src/components/Characters/Action'
import { ContextAppProvider } from '../../../src/context/AppProvider'
import { updateFavorites } from '../../../src/utils'

const MockContextProvider = ({ children }) => {
  const [favorites, setFavorites] = useState<number[]>([])

  const addFavorites = (id: number) => {
    const newFavorites = updateFavorites(favorites, id)
    setFavorites(newFavorites)
  }

  return (
    <ContextAppProvider.Provider
      value={{
        favorites,
        addFavorites,
        data: null,
        filter: '',
        setFiler: () => {},
        setData: () => {}
      }}
    >
      {children}
    </ContextAppProvider.Provider>
  )
}

describe('Pruebas en <Action/>', () => {
  const mockCharacterId = 1234
  const renderComponent = (idCharacter: number, characterDetails?: boolean) => {
    render(
      <MockContextProvider>
        <Action idCharacter={idCharacter} characterDetails={characterDetails} />
      </MockContextProvider>
    )
  }
  it('Debe cambiar el ícono cuando el usuario añada a favoritos el character', () => {
    renderComponent(mockCharacterId)
    

    const button = screen.getByTestId('action-component')

    expect(button.innerHTML).toContain('heart-black.svg')

    fireEvent.click(button)

    expect(button.innerHTML).toContain('heart.svg')
  })

  it('Debe cambiar el ícono cuando el usuario quite de favoritos el character', () => {
    renderComponent(mockCharacterId)

    const button = screen.getByTestId('action-component')
    fireEvent.click(button) // Agregar a favoritos
    fireEvent.click(button) // Quitar de favoritos

    expect(button.innerHTML).toContain('heart-black.svg')
  })

  it('Debe ajustar el tamaño del ícono según la propiedad characterDetails', () => {
    renderComponent(mockCharacterId, true)

    const buttonWithDetails = screen.getByTestId('action-component')
    const imgWithDetails = buttonWithDetails.querySelector('img')

    expect(imgWithDetails?.height).toBe(24)
    expect(imgWithDetails?.width).toBe(24)
  })

  it('Debe mantener el tamaño del ícono cuando characterDetails es false', () => {
    renderComponent(mockCharacterId, false);

    const buttonWithoutDetails = screen.getByTestId('action-component');
    const imgWithoutDetails = buttonWithoutDetails.querySelector('img');

    expect(imgWithoutDetails?.height).toBe(12);
    expect(imgWithoutDetails?.width).toBe(12);
  });
})
