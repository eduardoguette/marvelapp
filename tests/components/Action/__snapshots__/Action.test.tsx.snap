// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pruebas en <Action/> > Debe cambiar el ícono cuando el usuario añada a favoritos el character 1`] = `undefined`;
